skimage/filters/thresholding.py

```diff
@@ -719,3 +722,165 @@ def threshold_triangle(image, nbins=256):
+
+    Parameters
+    ----------
+    image : (N, M) ndarray
+        Input image.
+    w : int
+        Odd window size (e.g. 3, 5, 7, ..., 21, ...).
+
+    Returns
+    -------
+    m : 2-D array of same size of image with local mean values.
+    s : 2-D array of same size of image with local standard
+        deviation values.
+
+    References
+    ----------
+    .. [1] F. Shafait, D. Keysers, and T. M. Breuel, "Efficient
+           implementation of local adaptive thresholding techniques
+           using integral images." in Document Recognition and
+           Retrieval XV, (San Jose, USA), Jan. 2008.
+           DOI:10.1117/12.767755
+    """
+    from skimage.transform.integral import integral_image
+
+    if w == 1 or w % 2 == 0:
+        raise ValueError(
+            "Window size w = %s must be odd and greater than 1." % (w))
+    integrated = integral_image(image)  # Integral Image.
+
+    # Pad left and top of Integral image with zeros
+    integrated = np.pad(integrated, 1, mode='constant')[:-1,:-1]
+
+    kern = np.zeros((w + 1, w + 1))
+    kern[0, 0] = 1
+    kern[-1, -1] = 1
+    kern[[0, -1], [-1, 0]] = -1
+    # w2 holds total of pixels in window for each pixel (usually w x w).
+    w2 = ndi.convolve(
+        np.ones(image.shape, np.float), np.ones((w, w)), mode='constant')
+
+    m = ndi.convolve(integrated, kern, mode='nearest')[:-1, :-1] / w2
+    g = image.astype(np.float)
+    sum_g2 = ndi.convolve(g ** 2., np.ones((w, w)), mode='constant')
+    sum_m2 = w2 * m ** 2.
+    s = np.sqrt((sum_g2 - sum_m2) / w2)
+    return m, s
+
+
+def threshold_niblack(image, window_size=15, k=0.2):
+    """Applies Niblack local threshold to an array.
+
+    A threshold T is calculated for every pixel in the image using the
+    following formula:
+
+    T = m(x,y) - k * s(x,y)
+
+    where m(x,y) and s(x,y) are the mean and standard deviation of
+    pixel (x,y) neighborhood defined by a rectangular window with size w
+    times w centered around the pixel. k is a configurable parameter
+    that weights the effect of standard deviation.
+
+    Parameters
+    ----------
+    image: (N, M) ndarray
+        Grayscale input image.
+    window_size : int, optional
+        Odd size of pixel neighborhood window (e.g. 3, 5, 7...).
+    k : float, optional
+        Value of parameter k in threshold formula.
+
+    Returns
+    -------
+    threshold : (N, M) ndarray
+        Threshold mask. All pixels with an intensity higher than
+        this value are assumed to be foreground.
+
+    Notes
+    -----
+    This algorithm is originally designed for text recognition.
+
+    References
+    ----------
+    .. [1] Niblack, W (1986), An introduction to Digital Image
+           Processing, Prentice-Hall.
+
+    Examples
+    --------
+    >>> from skimage import data
+    >>> image = data.page()
+    >>> binary_image = threshold_niblack(image, window_size=7, k=0.1)
+    """
+    assert_nD(image, 2)
```

Comment:
As with the entropy PR, this can be very easily made n-dimensional, so please remove this check.